### PR TITLE
rt: add Handle::block_on

### LINKED ISSUE #2965: Replacement for Handle::current
With the removal of `Handle`, there is no equivalent to `Handle::current`. For example, the following is no longer easily possible:
```rust
let handle = Handle::current();
std::thread::spawn(move || {
    handle.block_on(...);
});
```

### PR DESCRIPTION
**Pull request by davidpdrsn**
This is a continuation of https://github.com/tokio-rs/tokio/pull/3549 but with tests for the behavior described [here](https://github.com/tokio-rs/tokio/pull/3549#issuecomment-785382050) and part of the implementation.

Still missing network operations failing if runtime associated with the handle has been shutdown. Changes to make timers and filesystem operations fail are at least passing the tests.

This is my first contribution to Tokio but let me know if I'm on the right track here 😊

Refs: #3097
Fixes #2965, #3096
cc @stuhood, @Keruspe

### CODE REVIEW

File: tokio/src/fs/mod.rs
Comment (Darksonn) [inline tokio/src/fs/mod.rs:115]: We want to keep all instances of these panic messages the same. See #3361.

Though in this case it probably should be panicking anyway inside the `sys::run` call. If it doesn't, fix `spawn_blocking` and this should fix itself.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:206]: Just adding this comment so we don't forget to fix doc.

File: tokio/src/time/driver/entry.rs
Comment (Darksonn) [inline tokio/src/time/driver/entry.rs:552]: Also #3361 here.

File: tokio/src/runtime/tests/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/tests/handle.rs:263]: Yes, calling `block_on` inside `block_on` should panic.

File: tokio/src/runtime/tests/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/tests/handle.rs:257]: Should this panic? It might just need to return an IO error.

**Review by Darksonn — COMMENTED** [review]
Some comments here.

File: tokio/src/fs/mod.rs
Comment (davidpdrsn) [reply tokio/src/fs/mod.rs:115]: Do you mean that these panics should use `CONTEXT_MISSING_ERROR` as the message or that there should be constants for each message in `tokio/src/util/error.rs`?

File: tokio/src/runtime/tests/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/tests/handle.rs]: 👍

File: tokio/src/runtime/tests/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/tests/handle.rs:257]: It shouldn't panic to be consistent with timers and file io?

File: tokio/src/runtime/blocking/pool.rs
Comment (LucioFranco) [inline tokio/src/runtime/blocking/pool.rs:264]: I wonder if we should use the `Spawner` to track the shutdown rather than the blocking pools spawner?

File: tokio/src/time/driver/entry.rs
Comment (LucioFranco) [inline tokio/src/time/driver/entry.rs:498]: Do we want to move this to some sort of `AtomicBool`?

File: tokio/src/runtime/blocking/pool.rs
Comment (davidpdrsn) [reply tokio/src/runtime/blocking/pool.rs:264]: Perhaps 🤷 There is already [`is_shutdown` on `Core`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L54) which is accessible through `Handle.spawner`. Is that what you have in mind?

I guess we could also add an `AtomicBool` to [`Shared`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L70) that gets set [here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L770).

File: tokio/src/time/driver/entry.rs
Comment (davidpdrsn) [reply tokio/src/time/driver/entry.rs:498]: Yeah I was wondering if this lock could be bad. The `is_shutdown` is defined [here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/mod.rs#L149) which is accessible through an [`Arc<Mutex<_>>` here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/handle.rs#L9), therefore the lock.

I guess we could move [`Inner.is_shutdown`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/mod.rs#L135) to [`Handle`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/time/driver/handle.rs#L7) and make it an `AtomicBool`. Is that what you have in mind?

File: tokio/src/fs/mod.rs
Comment (carllerche) [reply tokio/src/fs/mod.rs:115]: I don’t think this should panic. Hitting this is not due to a user bug. This probably should be an Err

**Comment by carllerche** [comment]
Generally I dont think attempting an op on a shutdown runtime should panic. Instead Err is better. That said... sleep is an interesting case because it does not return Result.

File: tokio/src/fs/mod.rs
Comment (carllerche) [reply tokio/src/fs/mod.rs:115]: I'm not 100% sure 😆

Comment (carllerche) [reply tokio/src/fs/mod.rs:115]: Ok, based on the Discord convo, FS ops will already error when the runtime shuts down (by returning `Err`). We should keep this behavior as is. We may want to look to see if the behavior is properly documented.

**Comment by davidpdrsn** [comment]
Summary from discussions on Discord:

- `Handle::block_on` should simply return the output of the future.
- Things that aren't runtime dependent shouldn't care about the runtime being shutdown. They should continue to run.
  - This includes `tokio::signal::ctrl_c` and channels.
- Functions such as `TcpStream::connect` which return a result today and require the runtime to work should return an error if the runtime gets shutdown. This includes stuff in `tokio::net`.
  - Currently they will fail if put onto one runtime, moved to a second, and the first is shutdown. We should reuse the same error for this case.
- Futures created by functions from `tokio::fs` have two possible cases
  - If the task has started before the runtime is shutdown it should continue (cannot cancel in-flight).
  - Any task that has not yet started should error when given to `Handle::block_on` as the `JoinHandle` should return an error when awaited.
  - Same goes for `Stdin`, `Stderr`, `Stdout` as they also run on a blocking threadpool.
- Timer futures should panic as they don't return a `Result` today and are dependent on the runtime.
  - There is some uncertainty about whether this is the current behavior so we should verify this with tests.

**Comment by carllerche** [comment]
@davidpdrsn what's the status of this? I think it was blocked on shutting down the timer properly. Is that issue tracked somewhere?

**Comment by davidpdrsn** [comment]
@carllerche I haven't had more time to work in it. Will spend some time on it this week.

**Comment by davidpdrsn** [comment]
I've pushed a bunch more changes. I think all the requirements should have tests and be working.

Still missing docs but I'll work on that if people think the implementation looks sound.

File: tokio/tests/rt_handle_block_on.rs
Comment (davidpdrsn) [inline tokio/tests/rt_handle_block_on.rs:331]: This error message was already there. I think it makes sense to change it to `A Tokio 1.x context was found, but it is being shutdown.` for consistency. Wondering if that would be breaking change though.

File: tokio/src/io/driver/registration.rs
Comment (LucioFranco) [inline tokio/src/io/driver/registration.rs:57]:
```suggestion
    /// `Interest`. `new_with_interest` should be crate::util::error::RUNTIME_SHUTTING_DOWN_ERRORsed over `new` when you need
```

Comment (davidpdrsn) [reply tokio/src/io/driver/registration.rs]: Seems I got vim'ed.

Fixed in f2e44f92d76e0

File: tokio/src/io/driver/registration.rs
Comment (jonhoo) [inline tokio/src/io/driver/registration.rs:242]: This seems a little unfortunate since `self.handle` also calls `self.handle.inner()`, so we end up upgrading inner twice.

File: tokio/src/runtime/handle.rs
Comment (jonhoo) [inline tokio/src/runtime/handle.rs:219]: Should this `cfg_` just be inside the `fn shutdown` around the `if let`?

File: tokio/src/time/driver/entry.rs
Comment (jonhoo) [inline tokio/src/time/driver/entry.rs:546]: Are we concerned about the additional `inner.upgrade()` inside `is_shutdown` on every timer poll here?

Comment (davidpdrsn) [reply tokio/src/time/driver/entry.rs:546]: The `inner.upgrade()` is only on the io driver. This call is in the time driver which ends here https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/time/driver/mod.rs#L452.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs:219]: I did try that at first but the macro takes [an `item`](https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/macros/cfg.rs#L65) so can't easily be put into functions afaik. It also checks [quite a few features](https://github.com/EmbarkStudios/tokio/blob/rt-handle-block-on/tokio/src/macros/cfg.rs#L64-L80) so didn't feel right copy pasting that.

File: tokio/src/io/driver/registration.rs
Comment (davidpdrsn) [reply tokio/src/io/driver/registration.rs]: Ah yeah good catch. Should be fixed in 33fd64ebe08ee.

**Review by jonhoo — APPROVED** [review]

**Comment by carllerche** [comment]
We need to slow down here. Anything that touches I/O shutdown needs to factor in #3481 or it will be racy. I will try to provide a deeper review shortly.

**Comment by carllerche** [comment]
Thanks for the detailed tests. I am going to spend more time with this PR and will comment asap.

**Comment by carllerche** [comment]
Sigh, I'm sorry, it seems like this PR is exposing more bugs in shutdown :grimacing:

File: tokio/src/io/driver/registration.rs
Comment (carllerche) [inline tokio/src/io/driver/registration.rs:243]: I was surprised that you needed to add this logic to make the tests pass. I did some digging and it appears that it is needed because the multi-threaded runtimes parker holds a strong ref to the I/O driver, which means that wakers hold strong refs to the driver. This goes back to #3481.

**Review by carllerche — CHANGES_REQUESTED** [review]
Thanks for doing this. I left some notes inline.

This is definitely a bit of a tangled problem that you are tackling. The issue is that there are known bugs with the shutdown process that you are hitting when trying to write tests. I love that you took the initiative to try to fix I/O shutdown as part of this PR. Unfortunately, the proposed solution will be racy when concurrently shutting down and registering a new resource. Issue #3481 includes a proposed solution that should avoid races.

Fixing I/O shutdown is out of scope of *this* PR, but you also wrote great tests that won't pass until the shutdown bugs are fixed. I see two paths forward.

a) We pause this PR until #3481 is fixed, then rebase and land this one.
b) We mark tests that cannot pass until #3481 is fixed with `#[ignore]` and reference the issue. Then we revert the `is_shutdown: AtomicBool` part of this PR and land it w/ known issues.

If I had to pick one, I would opt for `b` as we do need `Handle::block_on` and the shutdown bugs are unrelated. Thoughts?

**Comment by jonhoo** [comment]
I think I agree that b is the right path forward — we don't expect this to introduce _additional_ broken behavior, and we believe (I think) that we'd still want the same public API once we fix #3481.

**Comment by davidpdrsn** [comment]
Yeah I think option b makes sense. I have removed the `is_shutdown: AtomicBool` part and marked the tests that failed with `#[ignore]` and added a little note about why.

I will write some initial docs for `Handle::block_on` today.

**Comment by Darksonn** [comment]
If this doesn't introduce additional broken behavior, I'm fine with option b as well.

**Comment by davidpdrsn** [comment]
I've written some initial docs.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:213]: Is this true? Will it drive the IO driver when its the only call to block_on on a current_thread runtime?

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:220]: A runtime can also be shut down by dropping it.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:226]: Does it include signals? Assuming you refer to `tokio::signal`.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:233]: This doens't seem relevant to the `block_on` docs.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:251]: Maybe just remove this, replacing it with a blanket statement that things that require a Tokio runtime will fail if used during shutdown of the runtime.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:257]: This could be simplified to "this function panics if the provided future panics". We don't have to list a few examples of why it might panic.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:292]: Maybe this example would be better as an example that uses `#[tokio::main]`?

Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: Sure 👍

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: I personally consider these important details and something that should be explained in the docs.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: The first half was copied from `Runtime::block_on` so if its relevant there I would consider it relevant here as well. In general I prefer when docs are specific about when things panic as it helps users know what to expect.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: Good point! Added note about that.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs:213]: I remember checking that some time ago. But now I'm not sure. I'll have a look and perhaps add a test for it.

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [reply tokio/src/runtime/handle.rs]: Still, I would prefer saying something simpler like:

> If the runtime is shut down while a call to `block_on` is running, IO resources will return an error and timers will panic.

In particular, I don't think we should be explaining _why_ timers panic rather than return an error here.

Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: I think thats fine. Still gets the point across. I changed things in e1b6ebd3ce44e.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs:213]: Ah wait sorry. I commented on the wrong thing.

> Is this true? Will it drive the IO driver when its the only call to block_on on a current_thread runtime?

I'm not sure about this. Don't know enough about the internals to say 🤷 Do you know?

Comment (Darksonn) [reply tokio/src/runtime/handle.rs:213]: The `Handle::block_on` in Tokio 0.2 didn't do this, but I don't know if that has changed since then. To be clear, it is plausible that it could have changed as part of `Runtime::block_on` being changed from `&mut self` to `&self`. But I don't know.

@carllerche do you know?

Comment (carllerche) [reply tokio/src/runtime/handle.rs:213]: `Handle::block_on` and `Runtime::block_on` are slightly different. `Handle::block_on` **will not** drive the runtime where as `Runtime::block_on` **may** drive the runtime. Currently, this only happens with the `current_thread` scheduler.

File: tokio/src/runtime/handle.rs
Comment (carllerche) [inline tokio/src/runtime/handle.rs:273]: Nit: this does not need to be `_` prefixed (as it is used below).

**Review by carllerche — COMMENTED** [review]
LGTM :+1: There is the doc tweak + a minor style nit, but everything else looks good to me. I will let @Darksonn take it from here.

**Comment by carllerche** [comment]
It does look like CI needs to be fixed as well.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs]: Oh yeah of course. Fixed!

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs:213]: Alright. Is this something you think is worth mentioning or should I just cut out this bit?

Comment (Darksonn) [reply tokio/src/runtime/handle.rs:214]: Yes, we should mention this.

**Comment by jeromegn** [comment]
FWIW, we've been running this in production on all our servers for 2 days.

File: tokio/src/runtime/handle.rs
Comment (davidpdrsn) [reply tokio/src/runtime/handle.rs:214]: I've been thinking about this but sure how to word it accurately.

Comment (Darksonn) [reply tokio/src/runtime/handle.rs:214]: How about this?

> When this is used on a current_thread runtime, only the `Runtime::block_on` method can drive the IO and timer drivers, but the `Handle::block_on` method cannot drive them. This means that, when using this method on a current_thread runtime, anything that relies on IO or timers will not work unless there is another thread currently calling `Runtime::block_on` on the same runtime.

Comment (carllerche) [reply tokio/src/runtime/handle.rs:214]: I am applying ^^ to the PR so we can merge :)

Comment (carllerche) [reply tokio/src/runtime/handle.rs:214]: Actually, it seems like I can't push to your PR :'(

File: tokio/src/runtime/handle.rs
Comment (carllerche) [inline tokio/src/runtime/handle.rs:214]:
```suggestion
    /// When this is used on a `current_thread` runtime, only the
    /// [`Runtime::block_on`] method can drive the IO and timer drivers, but the
    /// `Handle::block_on` method cannot drive them. This means that, when using
    /// this method on a current_thread runtime, anything that relies on IO or
    /// timers will not work unless there is another thread currently calling
    /// [`Runtime::block_on`] on the same runtime.
```

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:261]: This example would panic.
```suggestion
    ///     let handle = Handle::current();
    ///     std::thread::spawn(move || {
    ///         // Using Handle::block_on to run async code in the new thread.
    ///         handle.block_on(async {
    ///             println!("hello");
    ///         });
    ///     });
```
I think it would be good to not have `no_run` on the examples.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**
